Add C++ end to end test for RPC, RRef and distributed autograd.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36893 Add C++ end to end test for RPC, RRef and distributed autograd.**

Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

pritamdamania committed:
Add C++ end to end test for RPC and distributed autograd.

Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania87 commented:
Will update the PR to also include RRef usage.

pritamdamania committed:
Update on "Add C++ end to end test for RPC and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli @jjlilley This PR is now ready to review, could you take a look? Thanks!

jjlilley commented on test/cpp/rpc/CMakeLists.txt:
```diff
@@ -2,15 +2,12 @@ set(TORCH_RPC_TEST_DIR "${TORCH_ROOT}/test/cpp/rpc")
   ${TORCH_ROOT}/test/cpp/common/main.cpp
   ${TORCH_RPC_TEST_DIR}/test_wire_serialization.cpp
-  ${TORCH_RPC_TEST_DIR}/test_tensorpipe_serialization.cpp
+  ${TORCH_RPC_TEST_DIR}/test_e2e_process_group.cpp
```
Can you double-check this, from when you were merging?
I think you probably didn't intend to remove the tensor pipe targets?

jjlilley commented on test/cpp/rpc/CMakeLists.txt:
```diff
@@ -2,15 +2,12 @@ set(TORCH_RPC_TEST_DIR "${TORCH_ROOT}/test/cpp/rpc")
-    ${ATen_CPU_INCLUDE}
-    ${TORCH_ROOT}/third_party/tensorpipe)
-target_link_libraries(test_cpp_rpc PRIVATE torch gtest)
+target_include_directories(test_cpp_rpc PRIVATE ${ATen_CPU_INCLUDE})
```
same here

mrshenli commented on test/cpp/rpc/e2e_test_base.h:
```diff
@@ -0,0 +1,152 @@
+      autograd::DistEngine::getInstance().execute(
+          DistAutogradContainer::currentContextId(),
+          {torch::sum(result)},
+          /* retainGraph */ false);
```
Do we also want to check the grad value?

mrshenli reviewed (changes requested):
This is great! Thanks for adding this.

Requesting changes because build failures are real.

```
May 07 18:40:44 FAILED: bin/test_cpp_rpc 
...
May 07 18:40:44 Undefined symbols for architecture x86_64:
May 07 18:40:44   "torch::distributed::rpc::TestE2EBase::numWorkers", referenced from:
May 07 18:40:44       torch::distributed::rpc::TestE2EBase::SetUp() in test_e2e_process_group.cpp.o
May 07 18:40:44       torch::distributed::rpc::TestE2EProcessGroup::buildRpcAgent() in test_e2e_process_group.cpp.o
May 07 18:40:44 ld: symbol(s) not found for architecture x86_64
```

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

pritamdamania87 commented:
@mrshenli Could you take another look at this PR? I had to do some major refactor to avoid python dependency in this unit test.

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]

mrshenli reviewed (commented):
New tests LGTM! I checked the log in https://circleci.com/api/v1.1/project/github/pytorch/pytorch/6128144/output/106/0?file=true&allocation-id=5f037a7d39d7392509bf833b-0-build%2F1AAB2550

but didn't see the log for `TestE2EProcessGroup`. Do you know if the tests are running in OSS? Other C++ tests seem not running, see #41143

mrshenli reviewed (commented):
Thanks for refactoring `request_callback_impl.cpp` code!!

mrshenli commented on torch/csrc/distributed/rpc/request_callback_no_python.cpp:
```diff
@@ -0,0 +1,625 @@
+      return;
+    }
+    case MessageType::PYTHON_CALL: {
+      processPythonCall(rpc, markComplete, messageId, responseFuture);
```
any reason for abstract some code into a dedicated method, but leaving others out (e.g., `SCRIPT_REMOTE_CALL`)?

pritamdamania87 replied:
I mostly focussed on refactoring the pieces that enable the python vs no python version. Didn't want to pile on too many more changes for this PR.

mrshenli reviewed (approved):
Stamp to unblock, assuming the new C++ tests did run internally, so that missing it in OSS CI is not a regression. 

For the code refactoring, there seems to be some inconsistency for different message types, but as this version is better than what we have currently, I think it is OK to merge.

pritamdamania committed:
Update on "Add C++ end to end test for RPC, RRef and distributed autograd."


Adding an end to end test for running a simple training loop in C++
for the distributed RPC framework.

The goal of this change is to enable LeakSanitizer and potentially catch memory
leaks in the Future. Enabling LSAN with python multiprocessing is tricky and we
haven't found a solution for this. As a result, adding a C++ test that triggers
most of the critical codepaths would be good for now.

As an example, this unit test would've caught the memory leak fixed by:
https://github.com/pytorch/pytorch/pull/31030

Differential Revision: [D21112208](https://our.internmc.facebook.com/intern/diff/D21112208/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D21112208/)!

[ghstack-poisoned]
